hive/replays/ppo_replay.py

```diff
@@ -0,0 +1,163 @@
+                (type, shape) tuple.
+            num_players_sharing_buffer (int): Number of agents that share their
+                buffers. It is used for self-play.
+        """
+        if extra_storage_types is None:
+            extra_storage_types = dict()
+        extra_storage_types.update(
+            {
+                "values": (np.float32, ()),
+                "returns": (np.float32, ()),
+                "advantages": (np.float32, ()),
+                "logprob": (np.float32, ()),
+            }
+        )
+        super().__init__(
+            capacity + stack_size - 1,
+            stack_size,
+            n_step,
+            gamma,
+            observation_shape,
+            observation_dtype,
+            action_shape,
+            action_dtype,
+            reward_shape,
+            reward_dtype,
+            extra_storage_types,
+            num_players_sharing_buffer,
+        )
+        self._use_gae = use_gae
+        self._gae_lambda = gae_lambda
+
+    # Taken from https://github.com/vwxyzjn/ppo-implementation-details/blob/main/ppo_shared.py
+    def compute_advantages(self, values):
+        """Compute advantages using rewards and value estimates."""
+        if self._use_gae:
+            last_gae_lambda = 0
+            for t in reversed(range(self._capacity)):
+                next_values = (
+                    values
+                    if t == self._capacity - 1
+                    else self._storage["values"][t + 1]
+                )
+                next_non_terminal = 1.0 - self._storage["done"][t]
+                delta = (
+                    self._storage["reward"][t]
+                    + self._gamma * next_values * next_non_terminal
+                    - self._storage["values"][t]
+                )
+                self._storage["advantages"][t] = last_gae_lambda = (
+                    delta
+                    + self._gamma
+                    * self._gae_lambda
+                    * next_non_terminal
+                    * last_gae_lambda
+                )
+            self._storage["returns"] = (
+                self._storage["advantages"] + self._storage["values"]
+            )
+        else:
+            for t in reversed(range(self._capacity)):
+                next_return = (
+                    values
+                    if t == self._capacity - 1
+                    else self._storage["return"][t + 1]
+                )
+                next_non_terminal = 1.0 - self._storage["done"][t]
+                self._storage["returns"][t] = (
+                    self._storage["reward"][t]
+                    + self._gamma * next_non_terminal * next_return
+                )
+            self._storage["advantages"] = (
+                self._storage["returns"] - self._storage["values"]
+            )
+
+    def reset(self):
+        """Resets the storage."""
+        if self._stack_size > 1:
+            transitions = {
```

Comment:
Maybe give a better name, something like saved_transitions? Or something else, so that the code is a bit more self-documenting.